crypto/src/kex/x25519.rs

```diff
@@ -43,30 +49,54 @@ impl KeyExchangeScheme for X25519Sha256 {
                 (pk, sk)
             }
         };
-        Ok((
-            PublicKey {
-                digest_function: ALGORITHM,
-                payload: ConstVec::new(pk.as_bytes().to_vec()),
-            },
-            PrivateKey {
-                digest_function: ALGORITHM,
-                payload: ConstVec::new(sk.to_bytes().to_vec()),
-            },
-        ))
+
+        let montgomery = curve25519_dalek::MontgomeryPoint(pk.to_bytes());
+        // 0 here means the positive sign, but it doesn't matter, because in
+        // `compute_shared_secret()` we convert it back to Montgomery form loosing the sign.
+        let edwards = montgomery
+            .to_edwards(0)
+            .expect("Montgomery to Edwards conversion failed");
+        let edwards_compressed = edwards.compress();
+
+        (
+            PublicKey::Ed25519(
+                crate::ed25519::PublicKey::from_bytes(edwards_compressed.as_bytes()).expect(
+                    "Ed25519 public key should be possible to create from X25519 public key",
+                ),
+            ),
+            PrivateKey::Ed25519(Box::new(crate::ed25519::PrivateKey::from_bytes(
+                sk.as_bytes(),
+            ))),
+        )
     }
 
     fn compute_shared_secret(
         &self,
         local_private_key: &PrivateKey,
         remote_public_key: &PublicKey,
-    ) -> SessionKey {
-        assert_eq!(local_private_key.digest_function, ALGORITHM);
-        assert_eq!(remote_public_key.digest_function, ALGORITHM);
-        let sk = StaticSecret::from(*array_ref!(&local_private_key.payload, 0, 32));
-        let pk = X25519PublicKey::from(*array_ref!(&remote_public_key.payload, 0, 32));
+    ) -> Result<SessionKey, Error> {
+        let crate::PrivateKey::Ed25519(local_private_key) = local_private_key else {
+            panic!("Wrong private key type, expected `Ed25519`, got {local_private_key:?}")
+        };
+        let crate::PublicKey::Ed25519(remote_public_key) = remote_public_key else {
+            panic!("Wrong public key type, expected `Ed25519`, got {remote_public_key:?}")
+        };
+
+        let sk = StaticSecret::from(*array_ref!(local_private_key.as_bytes(), 0, 32));
+
+        let pk_slice: &[u8; 32] = remote_public_key.as_bytes();
+        let edwards_compressed =
+            curve25519_dalek::edwards::CompressedEdwardsY::from_slice(pk_slice)
+                .expect("Ed25519 public key has 32 bytes");
+        let edwards = edwards_compressed.decompress().ok_or_else(|| {
+            Error::Parse("Invalid public key: failed to decompress edwards point".to_owned())
+        })?;
```

Comment:
Yes, we get it from another peer and we should not trust it

Reply:
shouldn't a thing like that be handled during deserialization? Once constructed `PublicKey` must be valid

Reply:
shouldn't this be verified during deserialization? Once constructed `PublicKey` must be valid

Reply:
The problem it that we store X25519 key as Ed25519 key, so not every valid Ed25519 is a valid X25519, see #4174 .

X25519 keys are used only for key exchaging and do not represent real keys used in Iroha. So it's a huge crutch. But this code is here for a long time and goes outside of my current scope because it requires a lot of fixes in downstream p2p code, so a created an issue.